### PR TITLE
add(api/enums): particle and sound effect enumerations

### PR DESCRIPTION
**Pull request by leagris**
- Add new GregTech API enumerations:
  - `ParticleFX`: Enumerate known EntityFX particles.
  - `SoundResource`: Enumerate known sounds with, id and ResourceLocation.
- Refactors code to use the new enumerations instead of string literals.
- Uses `ParticleFX` and `onRandomDisplayTick` to improve or implement
  new particle effects for these machines:
  - BBF: Adds random flames in front of the firebox.
  - Steam machines: Change pressure-exhaust particles to white vapour,
    rather than dark smoke.
  - Magic Energy Absorber: Adds random effect, of absorbed magical purple
    particles, by the EnderDragon Egg siphon.
  - Forge Hammer: Adds sparse random sparks, ejected from the main face.

### CODE REVIEW

File: src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java [hunk 1]
```diff
@@ -1,65 +1,59 @@
...
 import gregtech.api.objects.ItemData;
 import gregtech.api.render.TextureFactory;
-import gregtech.api.util.GT_AssemblyLineUtils;
-import gregtech.api.util.GT_Assemblyline_Server;
-import gregtech.api.util.GT_Log;
-import gregtech.api.util.GT_ModHandler;
-import gregtech.api.util.GT_OreDictUnificator;
-import gregtech.api.util.GT_Recipe;
-import gregtech.api.util.GT_Utility;
+import gregtech.api.util.*;
```
Comment (mitchej123) [inline src/main/java/gregtech/common/tileentities/machines/basic/GT_MetaTileEntity_Scanner.java:16]: nit: not  a fan of wildcard imports; I prefer them being explicit

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java [hunk 2]
```diff
@@ -24,64 +25,25 @@
...
-                    ofBlock(GregTech_API.sBlockCasings2, 0)
-            ))
-            .addElement('I', ofChain(
-                    // all blocks nearby use solid steel casing, so let's use the texture of that
-                    ofHatchAdder(GT_MetaTileEntity_AssemblyLine::addInputToMachineList, 16, 4),
-                    ofHatchAdder(GT_MetaTileEntity_AssemblyLine::addOutputToMachineList, 16,4)
-            ))
-            .addElement('i', ofHatchAdder(GT_MetaTileEntity_AssemblyLine::addInputToMachineList, 16, 5))
-            .build();
+    private static final IStructureDefinition<GT_MetaTileEntity_AssemblyLine> STRUCTURE_DEFINITION = StructureDefinition.<GT_MetaTileEntity_AssemblyLine>builder().addShape(STRUCTURE_PIECE_FIRST, transpose(new String[][]{{" ", "e", " "}, {"~", "l", "G"}, {"g", "m", "g"}, {"b", "i", "b"},})).addShape(STRUCTURE_PIECE_SECOND, transpose(new String[][]{{" ", "e", " "}, {"d", "l", "d"}, {"g", "m", "g"}, {"b", "I", "b"},})).addShape(STRUCTURE_PIECE_LATER, transpose(new String[][]{{" ", "e", " "}, {"d", "l", "d"}, {"g", "m", "g"}, {"b", "I", "b"},})).addElement('G', ofBlock(GregTech_API.sBlockCasings3, 10)) // grate machine casing
```
Comment (mitchej123) [inline src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java:40]: was this autoformatted?  It was more clear the way it was before

Comment (leagris) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java:40]: Fixed locally awayt pushing

Comment (leagris) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AssemblyLine.java]: Yes it was an unfortunate repeated <kbd>ctrl</kbd><kbd>shift</kbd><kbd>L</kbd> which reformats line-breaks in IntelliJ… and I had the nagging modal confirmation disabled.

**Review by mitchej123 — APPROVED** [review]

**Comment by Dream-Master** [comment]
@leagris can you resolve merge conflicts ?

**Pull request merged**